File: scripts/latest_build.rb
#!/usr/bin/env ruby

require 'net/http'

fmt = 'xx.xx , xx.xx.xx or xx.xx.x'
error = "Usage: eq_flush_script.rb <artifactory.username> \
      <artifactory.password> <service> <pomversion[optional]> \
      <pomversion> can be of the form #{fmt}"

unless ARGV.length >= 4 && ARGV.length <= 5
  puts error
  exit
end

services = ["actionsvc","actionexportersvc","casesvc","collectionexercisesvc", \
  "iacsvc","samplesvc","surveysvc","notifygatewaysvc","sdxgatewaysvc"]

unless services.include? ARGV[3]
  puts "Error! service <#{ARGV[3]}> not found!"
  exit
end

uname=ARGV[0]
passwd=ARGV[1]
host=ARGV[2]
service=ARGV[3]
version= nil
survey=false
if service.match(/surveysvc/)
  survey=true
end

if ARGV[4]
  if (ARGV[4].match(/^[0-9]{2}\.[0-9]{2}$/) || \
    ARGV[4].match(/^([0-9]{2}\.){2}[0-9]{1,2}$/)) || \
    ARGV[4].match(/^[0-9]{2,3}$/) && survey
    version = ARGV[4]
  else
    puts "#{ARGV[4]} is the wrong format, must be of the form #{fmt}."
    exit
  end
end

if survey
  ext = 'tar'
else
  ext = 'jar'
end

url="http://#{host}/artifactory/api/search/artifact?name=#{service}*#{version}*#{ext}&repos=libs-snapshot-local"

def createreq(uri)
  request = Net::HTTP::Get.new(uri)
  request.basic_auth $uname, $passwd
  return request
end

begin
  uri = URI(url)

  result = Net::HTTP.start(uri.hostname,80) {|http|
    http.request(createreq(uri))
  }
rescue SocketError
  puts "Host <#{host} not found!"
  exit
end

if result.body.match(/\"results\" \: \[ \]/)
  puts "Version #{version} not found for #{service}"
  exit
end

result.body.delete! "\"\}\]\s\n\{\["
result.body.gsub!('results:uri:','')
result.body.gsub!('api/storage/','')
snapshots = result.body.split("\,uri\:")

if survey
  result.body.gsub!(%r{[a-zA-Z\.\/\-\:]*},'')
  versions = result.body.split("\,")
  versions = versions.map(&:to_i)
  versions.sort!
  if version == nil
    version = versions[-1].to_s
    match = snapshots.each_index.select{|i| snapshots[i].match("-#{version}.")}
  else
    match = snapshots.each_index.select{|i| snapshots[i].match("-#{version.to_i}.")}
  end
  latest = snapshots[match[0].to_i]
else
  latest = snapshots[-1]
end

svc = "#{service}.#{ext}"

uri = URI(latest)
Net::HTTP.start(uri.hostname,80) do |http|
  puts "Downloading: #{latest}"
  resp = http.request(createreq(uri))
  open(svc, "wb") do |file|
      file.write(resp.body)
  end
end

File: scripts/LatestBuild.java
import java.util.*;
import java.io.*;
import java.net.*;
import java.nio.file.*;

public class LatestBuild {

  private static final String USER_AGENT = "Mozilla/5.0";
  private static final String SURVEYSVC = "surveysvc";
  private static final String[] SERVICES = {"actionsvc", "actionexportersvc",
                                      "casesvc", "collectionexercisesvc",
                                      "iacsvc", "samplesvc", "surveysvc",
                                      "notifygatewaysvc", "sdxgatewaysvc"};
  private static final String USAGE = "USAGE: java LatestBuild [-n] [-g] " +
                                      "<Host> <HomeDir> <Service>\n" +
                                      "-n   get name of latest version\n" +
                                      "-g   download latest version of " +
                                      "service\n";
  private static final int SERVICE_NAME_INDEX = 12;

  public static void main(String[] args) {

    String ext = "";
    String optn = "";
    String host = "";
    String home = "";
    String service = "";

    if (args.length == 4 || args.length == 5) {
      optn = args[0];
      host = args[1];
      home = args[2] + "/";
      service = args[3];
      if (args.length == 5 && optn.equals("-g")) {
        String sha = args[4];
        ext = ".git.sha" + sha;
      } else if (args.length == 4 && optn.equals("-n")) {
          if (service.equals(SURVEYSVC)) {
            ext = ".tar";
          }
          else ext = ".jar";
      } else {
        System.err.println(USAGE);
        System.exit(1);
      }
    } else {
      System.err.println(USAGE);
      System.exit(1);
    }

    if (!Arrays.asList(SERVICES).contains(service)) {
      System.err.println("Error! Service <" + service + "> does not exist!");
      System.exit(1);
    }

    String url = "" + host + "/artifactory/api/search/artifact?name=" +
       service + "*" + ext + "&repos=libs-snapshot-local";

    try {
      String snapshotList = getList(url, service, ext, home);
      String snapshotUrl = getSnapshotUrl(snapshotList, service);
      if (optn.equals("-g")) {
        getSvc(snapshotUrl, home);
      } else if (optn.equals("-n")) {
        getName(snapshotUrl, home, service);
      }
    } catch(Exception e) {
      e.printStackTrace();
    }
  }

  private static String getList(String uri, String svc, String ext, String home)
      throws IOException {

    HttpURLConnection connection = getConnection(uri);

    BufferedReader br = new BufferedReader(
      new InputStreamReader(connection.getInputStream()));

    StringBuffer response = new StringBuffer();
    String output;
    while ((output = br.readLine()) != null) {
      response.append(output);
    }

    br.close();
    connection.disconnect();
    return response.toString();
  }

  private static String getSnapshotUrl(String snapshotList, String service) {

    snapshotList = snapshotList.replaceAll("[\"\\} \\n\\{\\[\\]]*", "");
    snapshotList = snapshotList.replaceAll("results:uri:", "");
    snapshotList = snapshotList.replaceAll("api/storage/", "");

    List<String> snapshots = new ArrayList<String>(Arrays.asList(snapshotList
      .split("\\,uri\\:")));
    int urls = snapshots.size();
    String url = "";
    if (service.equals(SURVEYSVC)) {
      snapshotList = snapshotList.replaceAll("[a-zA-Z\\.\\/\\-\\:]*", "");
      List<String> surveyVersions = new ArrayList<String>(Arrays
        .asList(snapshotList.split("\\,")));

      List<Integer> versionsNo = new ArrayList<Integer>();
      for (String s : surveyVersions) {
        versionsNo.add(Integer.parseInt(s));
      }
      Collections.sort(versionsNo);
      int i = 0 ;
      for (Integer v : versionsNo) {
        surveyVersions.set(i, v.toString());
        i++;
      }
      i=0;
      for (String snapshot : snapshots) {
        if (snapshot.contains(surveyVersions.get(surveyVersions.size()-1))) {
          url = snapshot;
        }
        i++;
      }
    }
    else url = snapshots.get(urls-1);
    return url;
  }

  private static void getName(String snapshotUrl, String home, String svc)
      throws FileNotFoundException {

    String[] pathComponents = snapshotUrl.split("/");
    String svcVersion = pathComponents[SERVICE_NAME_INDEX]
      .replaceAll(".jar", ".git.sha");

    System.out.println(svcVersion);
  }

  private static void getSvc(String uri, String home)
      throws IOException {

    String url = uri.replaceAll("\\.git\\.sha.*$", ".jar");
    System.out.println(url);
    HttpURLConnection connection = getConnection(url);

    BufferedReader br = new BufferedReader(
      new InputStreamReader(connection.getInputStream()));
    String[] urlComponents = url.split("/");
    String filePath = home + urlComponents[12];
    Path path = Paths.get(filePath);
    Files.copy(connection.getInputStream(), path);
    br.close();
    connection.disconnect();
  }

  private static HttpURLConnection getConnection(String uri)
      throws MalformedURLException, IOException, ProtocolException {
    URL url = new URL(uri);
    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
    connection.setRequestMethod("GET");
    connection.setRequestProperty("User-Agent", USER_AGENT);

    return connection;
  }
}
